Diagnostics landing

sdmaclea opened the pull request:
## Summary

Add a diagnostic landing page and overview docs for using a debugger, logging and tracing.

Begins to fix #9276

noahfalk committed:
WIP - Add .NET Core diagnostics into material.

sdmaclea committed:
Add docs/core/diagnostics

Thraka commented:
Reviewing...

Thraka commented on docs/core/diagnostics/managed-debuggers.md:
```diff
@@ -0,0 +1,33 @@
+description: An overview of the Visual Studio and VS Code managed debuggers.
+author: stmaclea
+ms.author: stmaclea
+ms.date: 8/5/2019
```
```suggestion
ms.date: 08/05/2019
```

Thraka commented on docs/core/diagnostics/managed-debuggers.md:
```diff
@@ -0,0 +1,33 @@
+
+Debuggers allow programs to be paused or executed step-by-step. When paused, the current state of the process can be viewed. By stepping through key sections, you gain understanding of your code and why it produces the result that it does.
+
+Microsoft provides debuggers for managed code in the **Visual Studio** and **Visual Studio Code** integrated development environments.
```
```suggestion
Microsoft provides debuggers for managed code in the **Visual Studio** and **Visual Studio Code** integrated development environments (IDE).
```

Thraka commented on docs/core/diagnostics/managed-debuggers.md:
```diff
@@ -0,0 +1,33 @@
+
+## Visual Studio Managed Debugger
+
+**Visual Studio** is an IDE with the most comprehensive debugger available. Visual Studio is an excellent default choice for developers working on Windows.
```
I don't think **default** is really needed.

```suggestion
**Visual Studio** is an IDE with the most comprehensive debugger available. Visual Studio is an excellent choice for developers working on Windows.
```

Thraka commented on docs/core/diagnostics/managed-debuggers.md:
```diff
@@ -0,0 +1,33 @@
+## Visual Studio Managed Debugger
+
+**Visual Studio** is an IDE with the most comprehensive debugger available. Visual Studio is an excellent default choice for developers working on Windows.
+  - [Tutorial - Debugging a .NET Core application on Windows with Visual Studio](https://docs.microsoft.com/dotnet/core/tutorials/debugging-with-visual-studio)
```
Links should be relative to the current article and point to source files unless it points to a different repo.

```suggestion
  - [Tutorial - Debugging a .NET Core application on Windows with Visual Studio](../debugging-with-visual-studio.md)
```

Thraka commented on docs/core/diagnostics/managed-debuggers.md:
```diff
@@ -0,0 +1,33 @@
+
+ Debugging ASP.NET Core apps require slightly different instructions.
+
+  - [Debug ASP.NET Core apps in Visual Studio](https://docs.microsoft.com/visualstudio/debugger/how-to-enable-debugging-for-aspnet-applications#debug-aspnet-core-apps)
```
```suggestion
  - [Debug ASP.NET Core apps in Visual Studio](how-to-enable-debugging-for-aspnet-applications.md#debug-aspnet-core-apps)
```

sdmaclea replied:
This suggestion looks wrong.

Thraka replied:
Sorry, I saw the debugger folder and misread that as the current folder. Instead use `/visualstudio/....the-rest-of-your-link`

Thraka commented on docs/core/diagnostics/managed-debuggers.md:
```diff
@@ -0,0 +1,33 @@
+
+## Visual Studio Code Managed Debugger
+
+**Visual Studio Code** is a lightweight fully cross platform code editor. It uses the same .NET Core debugger implementation as Visual Studio with a simplified user interface.
```
```suggestion
**Visual Studio Code** is a lightweight cross-platform code editor. It uses the same .NET Core debugger implementation as Visual Studio, but with a simplified user interface.
```

Thraka commented on docs/core/diagnostics/index.md:
```diff
@@ -0,0 +1,27 @@
+description: An overview of the tools and APIs available to diagnose .NET Core applications.
+author: stmaclea
+ms.author: stmaclea
+ms.date: 8/5/2019
```
Date formats should be in MM/DD/YYYY

```suggestion
ms.date: 08/05/2019
```

Thraka commented on docs/core/diagnostics/index.md:
```diff
@@ -0,0 +1,27 @@
+ms.author: stmaclea
+ms.date: 8/5/2019
+ms.topic: article
+ms.prod: .net-core
```
This is set automatically based on where you article is in the repo.

```suggestion
```

Thraka commented on docs/core/diagnostics/index.md:
```diff
@@ -0,0 +1,27 @@
+author: stmaclea
+ms.author: stmaclea
+ms.date: 8/5/2019
+ms.topic: article
```
I believe this also defaults in the repo to `conceptual`

```suggestion
```

Thraka commented on docs/core/diagnostics/index.md:
```diff
@@ -0,0 +1,27 @@
+ms.topic: article
+ms.prod: .net-core
+---
+# .NET Core Diagnostics
```
Please review the guidelines for an overview article here: https://review.docs.microsoft.com/en-us/help/contribute/contribute-how-to-write-overview?branch=master and see if you can structure it accordingly. These are general guidelines since this is not an "Azure Service" as the guideline was really written for. But it's a nice template and we're trying to conform (where it makes sense) to it.

sdmaclea replied:
I'm not sure overview is the correct template....  It is intended as an index or local TOC.

I revised it to match the style of the https://docs.microsoft.com/en-us/dotnet/core/additional-tools/ overview.

Thraka replied:
As a writing org we're moving away from index/local toc style articles and to overviews. The additional-tools article is a carryover from the old way.

sdmaclea replied:
I spent a lot of time looking at this. The .NET Core Guide does not have any good overview articles.  It looks like the Machine Learning and Spark sections are more up to date and are better references.

I seriously considered removing the overview altogether, but I think it is a good place to provide some guidance.

I improved this a bit more by changing the article title and adding a customer intent comment.

We have a few more PRs which will touch this section and this overview. I think this is a good place to start.

Thraka commented on docs/core/diagnostics/logging-and-tracing.md:
```diff
@@ -0,0 +1,96 @@
+author: stmaclea
+ms.author: stmaclea
+ms.date: 8/5/2019
+ms.topic: article
```
Please update the metadata here with the same suggestions as the previous article. However, note that this isn't an overview article. You can delete `ms.topic` completely since we have it configured to default to `conceptual`.

```suggestion
```

Thraka commented on docs/core/diagnostics/logging-and-tracing.md:
```diff
@@ -0,0 +1,96 @@
+
+Logging and tracing is a primitive technique used since the early days of computers. It simply involves instrumenting an application to produce output to be consumed later.
+
+Logging and tracing are similar.  As used here we'll distinguish:
```
```suggestion
Logging and tracing are similar. As used here we'll distinguish:
```

Thraka commented on docs/core/diagnostics/managed-debuggers.md:
```diff
@@ -0,0 +1,33 @@
+
+ Debugging ASP.NET Core apps require slightly different instructions.
+
+  - [Debug ASP.NET Core apps in Visual Studio](https://docs.microsoft.com/visualstudio/debugger/how-to-enable-debugging-for-aspnet-applications#debug-aspnet-core-apps)
```
```suggestion
  - [Debug ASP.NET Core apps in Visual Studio](/visualstudio/debugger/how-to-enable-debugging-for-aspnet-applications#debug-aspnet-core-apps)
```

Sorry, I didn't notice this link was pointing to visual studio.

sdmaclea committed:
Review feedback

sdmaclea commented:
@thraka I have revised based on your feedback.

sdmaclea committed:
tutorials/with-visual-studio-code#debug

Thraka reviewed (approved):
@mairaw Do you want to review the location/toc since this is new content?

Thraka commented:
Thanks @sdmaclea this is great work!

mairaw commented on docs/core/diagnostics/logging-and-tracing.md:
```diff
@@ -0,0 +1,94 @@
+---
```
file names shouldn't have small words such as and (see https://review.docs.microsoft.com/en-us/help/contribute/file-path-guidelines?branch=master#name-your-files)

so please rename this to logging-tracing.md

mairaw commented on docs/core/diagnostics/index.md:
```diff
@@ -0,0 +1,22 @@
+---
+title: .NET Core Diagnostics
+description: An overview of the tools and APIs available to diagnose .NET Core applications.
+author: stmaclea
```
author is the GitHub alias and ms.author the MS alias, so here and in the other articles should be 
```suggestion
author: sdmaclea
```

mairaw commented:
Yes, thanks for pinging me @Thraka. I'll continue my review later today or tomorrow.

sdmaclea committed:
More feedback

noahfalk commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,94 @@
+  - Writes messages to Windows event log
+  - System administrators expect fatal application error messages to appear in the Windows event log.
+
+## Performance Considerations
```
I'm wondering if we need this section or if we should put it somewhere else? Most of the page described options available for logging and then here we are switching to some best practices for authoring logging.

mairaw replied:
```suggestion
## Performance considerations
```

noahfalk commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,94 @@
+  - Because it's the simplest approach, it's often used for ad-hoc temporary debugging. This debug code is often never checked in to source control.
+- <xref:System.Diagnostics.Trace?displayProperty=nameWithType>
+  - Only enabled when `TRACE` is defined.
+  - Writes to attached <xref:System.Diagnostics.TraceListener>s. Usually a debugger.
```
I'm not sure what the "Usually a debugger" means?

mairaw replied:
Don't make the plural by attaching the s because it's bad for loc. Define what the TraceListener is. Object?

```suggestion
  - Writes to attached <xref:System.Diagnostics.TraceListener> objects. Usually a debugger.
```

noahfalk commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,94 @@
+- <xref:System.Diagnostics.Trace?displayProperty=nameWithType>
+  - Only enabled when `TRACE` is defined.
+  - Writes to attached <xref:System.Diagnostics.TraceListener>s. Usually a debugger.
+  - It seems best to use this API when creating release-oriented logs.
```
I'm not sure how to interpret 'release-oriented'? Were you trying to call attention to the fact that DEBUG is usually only defined debug builds vs. TRACE is usually defined all builds?

mairaw replied:
It seems best is a weird recommendation. can we be more direct?

noahfalk reviewed (commented):
Thanks @sdmaclea, this seems like a great start. I made various suggestions and I'm still thinking about parts, but I hopefully its useful feedback.

noahfalk commented:
Adding various folks who might have an interest and good suggestions for these docs. @davidfowl @shirhatti @MarioHewardt @dotnet/dotnet-diag

sdmaclea commented on docs/core/diagnostics/index.md:
```diff
@@ -0,0 +1,22 @@
+This article helps you find the various tools you need.
+
+## [Managed Debuggers](managed-debuggers.md)
+Managed debuggers allow you to interact with your program. Debuggers give you insight into the behavior of your code.
```
I was struggling to keep this informal.  The next patch will try to keep some of the high level concepts yet keep the informal conversational style.

noahfalk replied:
I agree with the desire to keep it informal (or perhaps to use simple language). I was trying to balance that with providing enough detail that someone reading would have a hunch which link they should follow. Perhaps 'incrementally execute' -> 'step through', replace 'state' with variables or call stacks, and eliminate the 'flow' part?

sdmaclea committed:
Index feedback

sdmaclea commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,94 @@
+---
+# .NET Core Logging and Tracing
+
+Logging and tracing is a primitive technique used since the early days of computers. It simply involves instrumenting an application to produce output to be consumed later.
```
Reworked this slightly differently

Thraka commented on docs/toc.yml:
```diff
@@ -267,6 +267,13 @@
       href: core/docker/build-container.md
     - name: Container Tools in Visual Studio
       href: /visualstudio/containers/overview
+  - name: Diagnostics
```
Our site has changed the way these parent nodes operate. If the parent node, such as this one, has an article attached to it, it gets replicated as a child automatically. This causes the TOC to have two entries, one expandable and one clickable. They used to be expand+click but not anymore. The better way to do this is to create a true child with an alternative title:

```yml
  - name: Diagnostics
    items:
    - name: Overview
      href: core/diagnostics/index.md
    - name: Managed Debuggers
      href: core/diagnostics/managed-debuggers.md
    - name: Logging and Tracing
      href: core/diagnostics/logging-tracing.md
```

sdmaclea committed:
TOC feedback

sdmaclea committed:
More index feedback

Make index closer to an overview
Add customer intent
Clean up wording.

sdmaclea commented on docs/core/diagnostics/index.md:
```diff
@@ -0,0 +1,22 @@
+Logging and tracing are related techniques. They allow instrumenting code to create log files that can be used to diagnose the most complex problems.
+
+## [Unit Testing](../testing/index.md)
+Unit testing is a key component of continuous integration and deployment of high-quality software.
```
Over the course of the next month or two, I would like to fully populate this list.

mairaw commented:
There are a few broken links in the build report. I'll make suggestions on how to fix them.

mairaw commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,94 @@
+ms.author: stmaclea
+ms.date: 08/05/2019
+---
+# .NET Core Logging and Tracing
```
```suggestion
# .NET Core logging and tracing
```

mairaw commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,94 @@
+
+Logging and tracing is a primitive technique used since the early days of computers. It simply involves instrumenting an application to produce output to be consumed later.
+
+Logging and tracing are similar. As used here we'll distinguish:
```
We should be avoided unless Microsoft is recommending something
https://docs.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/w/we

```suggestion
Logging and tracing are similar. In this article, these two terms are distinguished as follows:

```

mairaw commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,94 @@
+
+Logging and tracing are similar. As used here we'll distinguish:
+- Logging outputs text as human readable content.
+- Tracing outputs binary content.  It can use tokenizing, and filtering to reduce overhead in the traced application.
```
```suggestion
- Tracing outputs binary content. It can use tokenizing and filtering to reduce overhead in the traced application.
```

mairaw replied:
tokenizing seems weird. Is that correct or should it be tokenization?

mairaw commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,94 @@
+- Logging outputs text as human readable content.
+- Tracing outputs binary content.  It can use tokenizing, and filtering to reduce overhead in the traced application.
+
+## Why Logging and Tracing?
```
```suggestion
## Reasons to use logging and tracing
```

mairaw commented on docs/core/diagnostics/index.md:
```diff
@@ -0,0 +1,23 @@
+---
+title: What Diagnostic Tools are Available in .NET Core?
+description: An overview of the tools and APIs available to diagnose .NET Core applications.
```
```suggestion
description: An overview of the tools and techniques available to diagnose .NET Core applications.
```

mairaw commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,94 @@
+
+## Logging APIs
+
+<xref:System.Console?displayProperty=nameWithType>, <xref:System.Diagnostics.Trace?displayProperty=nameWithType>, and <xref:System.Diagnostics.Debug?displayProperty=nameWithType> each provide similar APIs convenient for logging.
```
```suggestion
The <xref:System.Console?displayProperty=nameWithType>, <xref:System.Diagnostics.Trace?displayProperty=nameWithType>, and <xref:System.Diagnostics.Debug?displayProperty=nameWithType> classes provide similar APIs convenient for logging.
```

mairaw commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,94 @@
+
+<xref:System.Console?displayProperty=nameWithType>, <xref:System.Diagnostics.Trace?displayProperty=nameWithType>, and <xref:System.Diagnostics.Debug?displayProperty=nameWithType> each provide similar APIs convenient for logging.
+
+The choice of which API to use is up to you. The key differences:
```
```suggestion
The choice of which API to use is up to you. The key differences are:
```

mairaw commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,94 @@
+
+The choice of which API to use is up to you. The key differences:
+- <xref:System.Console?displayProperty=nameWithType>
+  - Always compiled in and always writes to the console.
```
what does compiled in means?

mairaw commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,94 @@
+The choice of which API to use is up to you. The key differences:
+- <xref:System.Console?displayProperty=nameWithType>
+  - Always compiled in and always writes to the console.
+  - Useful for information that your customer may need to see in release.
```
```suggestion
  - Useful for information that your customer may need to see in the release.
```

mairaw commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,94 @@
+  - It seems best to use this API when creating release-oriented logs.
+- <xref:System.Diagnostics.Debug?displayProperty=nameWithType>
+  - Only enabled when `DEBUG` is defined.
+  - Writes to attached <xref:System.Diagnostics.TraceListener>s. Usually a debugger.
```
Ditto

mairaw commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,94 @@
+- <xref:System.Diagnostics.Debug?displayProperty=nameWithType>
+  - Only enabled when `DEBUG` is defined.
+  - Writes to attached <xref:System.Diagnostics.TraceListener>s. Usually a debugger.
+  - It seems best to use this API when creating debug-oriented logs.
```
Ditto

mairaw commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,94 @@
+  - Writes to attached <xref:System.Diagnostics.TraceListener>s. Usually a debugger.
+  - It seems best to use this API when creating debug-oriented logs.
+
+### Logging Related References
```
```suggestion
### Logging related references
```

mairaw commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,94 @@
+
+### Logging Related References
+
+- [How to: Compile Conditionally with Trace and Debug](../../framework/debug-trace-profile/how-to-compile-conditionally-with-trace-and-debug)
```
```suggestion
- [How to: Compile Conditionally with Trace and Debug](../../framework/debug-trace-profile/how-to-compile-conditionally-with-trace-and-debug.md)
```

mairaw commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,94 @@
+
+- [How to: Compile Conditionally with Trace and Debug](../../framework/debug-trace-profile/how-to-compile-conditionally-with-trace-and-debug)
+
+- [How to: Add Trace Statements to Application Code](../../framework/debug-trace-profile/how-to-add-trace-statements-to-application-code)
```
```suggestion
- [How to: Add Trace Statements to Application Code](../../framework/debug-trace-profile/how-to-add-trace-statements-to-application-code.md)
```

mairaw commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,94 @@
+- [ASP.NET Logging](/aspnet/core/fundamentals/logging)
+provides an overview of the logging techniques it supports.
+
+- [C# String Interpolation](../../csharp/language-reference/tokens/interpolated)
```
```suggestion
- [C# String Interpolation](../../csharp/language-reference/tokens/interpolated.md)
```

mairaw commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,94 @@
+- [C# String Interpolation](../../csharp/language-reference/tokens/interpolated)
+ can simplify writing logging code.
+
+- The <xref:System.Exception.Message?displayProperty=nameWithType> is useful for logging exceptions.
```
```suggestion
- The <xref:System.Exception.Message?displayProperty=nameWithType> property is useful for logging exceptions.
```

mairaw commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,94 @@
+
+- The <xref:System.Exception.Message?displayProperty=nameWithType> is useful for logging exceptions.
+
+- The <xref:System.Diagnostics.StackTrace?displayProperty=nameWithType>
```
```suggestion
- The <xref:System.Diagnostics.StackTrace?displayProperty=nameWithType> class
```

mairaw commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,94 @@
+- The <xref:System.Exception.Message?displayProperty=nameWithType> is useful for logging exceptions.
+
+- The <xref:System.Diagnostics.StackTrace?displayProperty=nameWithType>
+can be useful to provide debugger like logging info.
```
```suggestion
can be useful to provide logging information to the debugger.
```

mairaw commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,94 @@
+## Tracing APIs
+
+- <xref:System.Diagnostics.Tracing.EventSource?displayProperty=nameWithType>
+  - EventSource is the primary root .NET Core tracing API
```
```suggestion
  - EventSource is the primary root .NET Core tracing API.
```

mairaw commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,94 @@
+
+- <xref:System.Diagnostics.Tracing.EventSource?displayProperty=nameWithType>
+  - EventSource is the primary root .NET Core tracing API
+  - Available in all .NET Standard versions
```
```suggestion
  - Available in all .NET Standard versions.
```

mairaw commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,94 @@
+- <xref:System.Diagnostics.Tracing.EventSource?displayProperty=nameWithType>
+  - EventSource is the primary root .NET Core tracing API
+  - Available in all .NET Standard versions
+  - Emits to .NET Core's event pipe on all platforms
```
```suggestion
  - Emits to .NET Core's event pipe on all platforms.
```

mairaw commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,94 @@
+  - EventSource is the primary root .NET Core tracing API
+  - Available in all .NET Standard versions
+  - Emits to .NET Core's event pipe on all platforms
+  - Also emits to ETW on Windows.
```
Yes, we need to spell out ETW. Link also helps.

mairaw commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,94 @@
+  - Available in all .NET Standard versions
+  - Emits to .NET Core's event pipe on all platforms
+  - Also emits to ETW on Windows.
+  - Only allows tracing serializable objects
```
```suggestion
  - Only allows tracing serializable objects.
```

mairaw commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,94 @@
+  - Only allows tracing serializable objects
+
+- <xref:System.Diagnostics.DiagnosticSource?displayProperty=nameWithType>
+  - Only available in .NET Core
```
```suggestion
  - Only available in .NET Core.
```

mairaw commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,94 @@
+
+- <xref:System.Diagnostics.DiagnosticSource?displayProperty=nameWithType>
+  - Only available in .NET Core
+  - Allows tracing non-serializable objects
```
```suggestion
  - Allows tracing non-serializable objects.
```

mairaw commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,94 @@
+  - Includes a bridge to allow select fields of logged objects to be written to an <xref:System.Diagnostics.Tracing.EventSource>.
+
+- <xref:System.Diagnostics.TraceSource?displayProperty=nameWithType>
+  - This API was introduced in .NET Framework 2.0
```
```suggestion
  - This API was introduced in .NET Framework 2.0.
```

mairaw replied:
Is this information relevant since this is a .NET Core article?

mairaw commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,94 @@
+
+- <xref:System.Diagnostics.TraceSource?displayProperty=nameWithType>
+  - This API was introduced in .NET Framework 2.0
+  - Supported in .NET Standard 2.0
```
```suggestion
  - Supported in .NET Standard 2.0.
```

mairaw commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,94 @@
+- <xref:System.Diagnostics.TraceSource?displayProperty=nameWithType>
+  - This API was introduced in .NET Framework 2.0
+  - Supported in .NET Standard 2.0
+  - Prefer <xref:System.Diagnostics.Tracing.EventSource>
```
```suggestion
  - Prefer <xref:System.Diagnostics.Tracing.EventSource> instead.
```

mairaw replied:
Do we want to say why?

sdmaclea replied:
On Noah's suggestion, I removed the `TraceSource` reference.

mairaw commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,94 @@
+  - Prefer <xref:System.Diagnostics.Tracing.EventSource>
+
+- <xref:System.Diagnostics.EventLog?displayProperty=nameWithType>
+  - Windows only
```
```suggestion
  - Windows only.
```

mairaw commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,94 @@
+
+- <xref:System.Diagnostics.EventLog?displayProperty=nameWithType>
+  - Windows only
+  - Writes messages to Windows event log
```
```suggestion
  - Writes messages to the Windows event log.
```

mairaw commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,94 @@
+
+String formatting can take noticeable CPU processing time.
+
+In performance critical applications, it's good to:
```
```suggestion
In performance critical applications, it's recommended that you:
```

mairaw commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,94 @@
+String formatting can take noticeable CPU processing time.
+
+In performance critical applications, it's good to:
+- Avoid lots of logging when no one is listening
```
```suggestion
- Avoid lots of logging when no one is listening.
```

mairaw commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,94 @@
+
+In performance critical applications, it's good to:
+- Avoid lots of logging when no one is listening
+- Only log what is useful
```
```suggestion
- Only log what's useful.
```

mairaw commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,94 @@
+In performance critical applications, it's good to:
+- Avoid lots of logging when no one is listening
+- Only log what is useful
+- Defer fancy formatting to the analysis stage
```
```suggestion
- Defer fancy formatting to the analysis stage.
```

mairaw commented on docs/core/diagnostics/managed-debuggers.md:
```diff
@@ -0,0 +1,30 @@
+---
+title: .NET Core Managed Debuggers
```
```suggestion
title: Managed debuggers - .NET Core
```

mairaw commented on docs/core/diagnostics/managed-debuggers.md:
```diff
@@ -0,0 +1,30 @@
+---
+title: .NET Core Managed Debuggers
+description: An overview of the Visual Studio and VS Code managed debuggers.
```
```suggestion
description: An overview of the Visual Studio and Visual Studio Code managed debuggers.
```

mairaw commented on docs/core/diagnostics/managed-debuggers.md:
```diff
@@ -0,0 +1,30 @@
+ms.author: stmaclea
+ms.date: 08/05/2019
+---
+# .NET Core Managed Debuggers
```
```suggestion
# .NET Core managed debuggers
```

mairaw commented on docs/core/diagnostics/managed-debuggers.md:
```diff
@@ -0,0 +1,30 @@
+
+Debuggers allow programs to be paused or executed step-by-step. When paused, the current state of the process can be viewed. By stepping through key sections, you gain understanding of your code and why it produces the result that it does.
+
+Microsoft provides debuggers for managed code in the **Visual Studio** and **Visual Studio Code** integrated development environments (IDE).
```
```suggestion
Microsoft provides debuggers for managed code in **Visual Studio** and **Visual Studio Code**.
```

mairaw replied:
What about Visual Studio for Mac?

Thraka commented:
His bad links are missing the `.md` extension

mairaw commented on docs/core/diagnostics/managed-debuggers.md:
```diff
@@ -0,0 +1,30 @@
+
+## Visual Studio Managed Debugger
+
+**Visual Studio** is an IDE with the most comprehensive debugger available. Visual Studio is an excellent choice for developers working on Windows.
```
```suggestion
**Visual Studio** is an integrated development environment with the most comprehensive debugger available. Visual Studio is an excellent choice for developers working on Windows.
```

mairaw commented on docs/core/diagnostics/managed-debuggers.md:
```diff
@@ -0,0 +1,30 @@
+**Visual Studio** is an IDE with the most comprehensive debugger available. Visual Studio is an excellent choice for developers working on Windows.
+- [Tutorial - Debugging a .NET Core application on Windows with Visual Studio](../tutorials/debugging-with-visual-studio)
+
+While Visual Studio is a Windows application, it can still be used to debug Linux and MacOS apps remotely.
```
```suggestion
While Visual Studio is a Windows application, it can still be used to debug Linux and macOS apps remotely.

```

mairaw commented on docs/core/diagnostics/managed-debuggers.md:
```diff
@@ -0,0 +1,30 @@
+- [Tutorial - Debugging a .NET Core application on Windows with Visual Studio](../tutorials/debugging-with-visual-studio)
+
+While Visual Studio is a Windows application, it can still be used to debug Linux and MacOS apps remotely.
+- [Debugging a .Net Core application on Linux/OSX with Visual Studio](https://github.com/Microsoft/MIEngine/wiki/Offroad-Debugging-of-.NET-Core-on-Linux---OSX-from-Visual-Studio)
```
```suggestion
- [Debugging a .NET Core application on Linux/OSX with Visual Studio](https://github.com/Microsoft/MIEngine/wiki/Offroad-Debugging-of-.NET-Core-on-Linux---OSX-from-Visual-Studio)
```

mairaw commented on docs/core/diagnostics/managed-debuggers.md:
```diff
@@ -0,0 +1,30 @@
+
+- [Debug ASP.NET Core apps in Visual Studio](/visualstudio/debugger/how-to-enable-debugging-for-aspnet-applications#debug-aspnet-core-apps)
+
+## Visual Studio Code Managed Debugger
```
```suggestion
## Visual Studio Code managed debugger
```

mairaw commented on docs/core/diagnostics/managed-debuggers.md:
```diff
@@ -0,0 +1,30 @@
+
+## Visual Studio Code Managed Debugger
+
+**Visual Studio Code** is a lightweight cross-platform code editor. It uses the same .NET Core debugger implementation as Visual Studio, but with a simplified user interface.
```
```suggestion
**Visual Studio Code** is a lightweight cross-platform code editor. It uses the same .NET Core debugger implementation as Visual Studio, but with a simplified user interface.

```

mairaw commented on docs/core/diagnostics/managed-debuggers.md:
```diff
@@ -0,0 +1,30 @@
+## Visual Studio Code Managed Debugger
+
+**Visual Studio Code** is a lightweight cross-platform code editor. It uses the same .NET Core debugger implementation as Visual Studio, but with a simplified user interface.
+- [Tutorial - Debugging a .Net Core application with Visual Studio Code](../tutorials/with-visual-studio-code#debug)
```
```suggestion
- [Tutorial - Debugging a .NET Core application with Visual Studio Code](../tutorials/with-visual-studio-code.md#debug)
```

mairaw commented on docs/core/diagnostics/managed-debuggers.md:
```diff
@@ -0,0 +1,30 @@
+
+**Visual Studio Code** is a lightweight cross-platform code editor. It uses the same .NET Core debugger implementation as Visual Studio, but with a simplified user interface.
+- [Tutorial - Debugging a .Net Core application with Visual Studio Code](../tutorials/with-visual-studio-code#debug)
+- [Visual Studio Code's Debugging Overview](https://code.visualstudio.com/docs/editor/debugging)
```
```suggestion
- [Debugging in Visual Studio Code](https://code.visualstudio.com/docs/editor/debugging)
```

mairaw commented on docs/core/diagnostics/managed-debuggers.md:
```diff
@@ -0,0 +1,30 @@
+
+Microsoft provides debuggers for managed code in the **Visual Studio** and **Visual Studio Code** integrated development environments (IDE).
+
+## Visual Studio Managed Debugger
```
```suggestion
## Visual Studio managed debugger
```

mairaw commented on docs/core/diagnostics/index.md:
```diff
@@ -0,0 +1,23 @@
+
+This article helps you find the various tools you need.
+
+## [Managed Debuggers](managed-debuggers.md)
```
```suggestion
## [Managed debuggers](managed-debuggers.md)
```

mairaw reviewed (changes requested):
Looks good @sdmaclea. Left some comments for you to consider. You can batch them into a single commit via the Files changed tab, if that makes it easier.

mairaw reviewed (changes requested):
Looks good @sdmaclea. Left some comments for you to consider. You can batch them into a single commit via the Files changed tab, if that makes it easier.

sdmaclea committed:
Revise logging doc

Incorporate feedback from Noah and Andrew

sdmaclea committed:
Apply Mairaw's suggestions

sdmaclea committed:
More feedback

hoyosjs commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,97 @@
+- <xref:System.Diagnostics.Trace?displayProperty=nameWithType>
+  - Only enabled when `TRACE` is defined.
+  - Writes to attached <xref:System.Diagnostics.Trace.Listeners>, by default the <xref:System.Diagnostics.DefaultTraceListener>.
+  - It seems best to use this API when creating logs that will be enabled in most builds.
```
Not really the type of language I'd expect to see in a guidance doc. As we are open to other vendors in this space, you can just say it's the preferred inbox solution for logging in production.

hoyosjs commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,97 @@
+  - Only enabled when `DEBUG` is defined.
+  - Writes to an attached debugger.
+  - On `*nix` writes to stderr if `COMPlus_DebugWriteToStdErr` is set.
+  - It seems best to use this API when creating logs that will only be useful in a debguenabled in debug builds.
```
Just as last comment, seems is a little vague. Maybe say: This API is useful to provide additional logging that's useful only if a [managed debugger](md link here) is attached. Also typo in `debguenabled`, although not sure if you mean `that will only be useful in debug-enabled builds`?

sdmaclea commented:
> What about Visual Studio for Mac?

Certainly this was an oversight on my part.  However as I look for relevant documentation to cite, I am struggling.  `.NET Core` support is barely documented anywhere for `Visual Studio for Mac`.

[The main Visual Studio for Mac page](https://docs.microsoft.com/visualstudio/mac/?view=vsmac-2019) links to [Debugging with Xamarin](https://docs.microsoft.com/visualstudio/mac/debugging?view=vsmac-2019) which very clearly talks about the mono debugger and its integration into the Mono runtime. It clearly implies no .NET Core debugger support.

[Visual Studio for Mac 2019 Tour](https://docs.microsoft.com/en-us/visualstudio/mac/ide-tour?view=vsmac-2019#debugging) uses similar language.

There is a mention of support in the `.NET Core Guide` [tutorial using-on-mac-vs-full-solution#adding-a-console-app item 6...](https://docs.microsoft.com/en-us/dotnet/core/tutorials/using-on-mac-vs-full-solution#adding-a-console-app), but it is underwhelming.

I am going to call `Visual Studio for Mac` `.NET Core` debugger support out of scope for this initial diagnostics PR. I will open a separate issue.

sywhang commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,97 @@
+  - Only allows tracing serializable objects.
+  - Writes to attached <xref:System.Diagnostics.Tracing.EventListener>s.
+  - .NET Core provides listeners for:
+    - .NET Core's event pipe on all platforms
```
I'd prefer to reword `event pipe`  to `EventPipe` for the sake of consistency across documentations.

sdmaclea commented:
Other than the Visual Studio for Mac debugger documentation, I think I have addressed all the feedback. This should be ready for a final review/merge.

sywhang commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,97 @@
+
+- <xref:System.Diagnostics.EventLog?displayProperty=nameWithType>
+  - Windows only.
+  - Writes messages to Windows event log.
```
nit: I think `Windows Event Log` is the name they use in Windows API docs so we might want to capitalize those E and L to match with their docs: https://docs.microsoft.com/en-us/windows/win32/wes/windows-event-log

sywhang commented on docs/core/diagnostics/logging-tracing.md:
```diff
@@ -0,0 +1,97 @@
+- <xref:System.Diagnostics.EventLog?displayProperty=nameWithType>
+  - Windows only.
+  - Writes messages to Windows event log.
+  - System administrators expect fatal application error messages to appear in the Windows event log.
```
same as above

sdmaclea committed:
Split the logging .NET Core APIs section

sdmaclea committed:
Correct Window Event Log capitalization

sdmaclea committed:
Use EventPipe spelling

sdmaclea commented:
@mairaw ping.

mairaw committed:
Update index.md

mairaw committed:
Update logging-tracing.md

mairaw reviewed (approved):
Looks good @sdmaclea. I thought it would be quicker if I just made some changes myself. Check if they look good to you.

sdmaclea reviewed (commented):
@mairaw Thanks.  Your changes look good to me.
